### DISCUSSION TITLE
Performance of Modbus RTU server

### DISCUSSION THREAD

**Discussion by MinzaGayas** [pymodbus-dev/pymodbus#942] [Q&A] [answered]
I am developing a Modbus server using pymodbus v2.5.3 synchronous server and RTU framer. The client is sending requests @600ms. I have set the serial socket with timeout of 1s because below this, framer does not send proper responses. Can the framer work at below 1s limit because it loses half of the requests sent by client?

**Comment by janiversen** [comment, marked as answer]
The framer does not have demands on timing, apart from the 3.5character wait between messages. The framer depends on the driver to do the read timeout.

**Reply by MinzaGayas** [reply]
I am running code on raspberry pi CM 4 running on Linux 11 bullseye OS. Is this the limitation at the hardware level i.e serial driver?

**Reply by janiversen** [reply]
The 3.5char between messages is part of the modbus protocol description. There are no protocol reason to have a 1s delay, but remember you do not run async, so you need to allow for task switching.

**Reply by MinzaGayas** [reply]
Baudrate is set at 9600 so 3.5 char delay in this case is not that much..yes I am using sync server and there is only one client sending sequential requests. If I reduce timeout in serial socket to 0, I  find frames are not having proper start so the client doesn't accept the response. Is there any limit on how fast a sync server can respond? Or could it be fixed at serial module level?

**Reply by janiversen** [reply]
0 is very tough, it means no delays at all are allowed, it’s natural it causes problems.

Of course there is a limit, but it depends on a lot of factors, starting how fast your serial driver can send/receive the data to pymodbus ( Not the client, which is defined by the baudrate) so I cannot tell you an exact number.

**Reply by MinzaGayas** [reply]
Thanks! I will focus on the driver then.

**Reply by MinzaGayas** [reply]
I have set timeout to 500ms and after logging the data, it is evident that requests from the client are received within 500ms limit but then framer on the server side gives frame check failed error or framer not ready on alternate packets. What could be the issue?

<img width="859" alt="rtu_framer" src="https://user-images.githubusercontent.com/104010847/174588719-d7161c7b-3eea-435b-9306-92cb63264090.PNG">

**Reply by janiversen** [reply]
500ms delay is (unless you have a very low bps) more than the 3.5char wait between messages so you get messages mixed up.

**Reply by MinzaGayas** [reply]
Baudrate for this serial connection is 9600 and 3.5 char wait is 4ms approx. for this rate. Since, the inverter sends requests@660ms, I want server to receive all the requests and I have set timeout to 500ms. How much should be this timeout delay be in this case? It is more than enough for 3.5char wait. Do I need to use lesser timeout? Because setting timeout to 0 or 1s, gives the same errors at the framer level.

**Reply by janiversen** [reply]
Timeout should normally be:
0 < timeout < time to send 3.5char

**Reply by MinzaGayas** [reply]
Does it also depend on the number of bytes serial socket has to read? say 256bytes in this case

**Reply by janiversen** [reply]
it is timeout, not a time to read.

**Reply by MinzaGayas** [reply]
I am trying to set this timeout of serial-Serial socket in _connect function of server file sync.py
`self.socket = serial.Serial(port=self.device,
                                        timeout=self.timeout,
                                        bytesize=self.bytesize,
                                        stopbits=self.stopbits,
                                        baudrate=self.baudrate,
                                        parity=self.parity)`
I have not made any changes to framer timings for 3.5char wait. 
With timeout set at 3ms, framer still gives these errors. Also, Modbus sniffer does not show the responses with a proper start if timeout is set less than 1s.
<img width="841" alt="bat_log_timeout500ms" src="https://user-images.githubusercontent.com/104010847/174604391-0d54a02c-e696-4147-bc0c-40bb4ebccb4f.PNG">

<img width="860" alt="rtu_framer_3ms" src="https://user-images.githubusercontent.com/104010847/174604202-38b644d5-2071-43d3-aaac-8c9843f52d22.PNG">

**Reply by MinzaGayas** [reply]
Hi @janiversen, 
After doing more analysis, I suppressed any responses from the server and just tried to receive client requests. Regardless of timeout value set, the framer either shows frame not ready or frame check failed for half of the requests being received. 
what could be the possible reason behind this? It does not seem to depend on serial read timeout at all. Is the client too fast for it?
<img width="860" alt="framer_check_error" src="https://user-images.githubusercontent.com/104010847/174760893-87e0d1bc-7090-4a3c-9154-1366267d0a56.PNG">

**Comment by janiversen** [comment]
How is your development progressing ?

**Reply by MinzaGayas** [reply]
It is still stuck with RTU framer not responding to below 1s request rate :(

**Reply by janiversen** [reply]
have you upgraded to our current version, a lot of changes went into version 3.0 and more in the other versions.

I have no problem with sending frames very fast.
